core/os/android/profiling/cc/launch_producer.cc

```diff
@@ -0,0 +1,154 @@
+  do {                                                      \
+    printf(name ": " msg "\n", ## __VA_ARGS__);             \
+    __android_log_print(lvl, "GAPID", msg, ## __VA_ARGS__); \
+  } while (false)
+
+#define LOG_ERR(msg, ...)  _LOG(ANDROID_LOG_ERROR, "E", msg, ## __VA_ARGS__)
+#define LOG_WARN(msg, ...) _LOG(ANDROID_LOG_WARN,  "W", msg, ## __VA_ARGS__)
+#define LOG_INFO(msg, ...) _LOG(ANDROID_LOG_INFO,  "I", msg, ## __VA_ARGS__)
+
+#define GET_PROP(name, trans)                       \
+  do {                                              \
+    char _v[PROP_VALUE_MAX] = {0};                  \
+    if (__system_property_get(name, _v) == 0) {     \
+      LOG_ERR("Failed reading property %s", name);  \
+      std::abort();                                 \
+    }                                               \
+    trans;                                          \
+  } while (0)
+#define GET_STRING_PROP(n, t) GET_PROP(n, t = _v)
+
+#if defined(__arm__)
+  #define ABI "armeabi-v7a"
+#elif defined(__aarch64__)
+  #define ABI "arm64-v8a"
+#elif defined(__i686__)
+  #define ABI "x86"
+#elif defined(__x86_64__)
+  #define ABI "x86_64"
+#else
+  #error "Unsupported ABI"
+#endif
+
+#define NELEM(x) (sizeof(x) / sizeof(x[0]))
+
+typedef void (*FN_PTR)(void);
+
+const char* kDriverProperty = "ro.gfx.driver.1";
+const char* kProducerPaths[] = {
+  "/assets/libgpudataproducer.so",
+  "/lib/" ABI "/libgpudataproducer.so",
+};
+
+std::string getDriver() {
+  std::string driver;
+
+  GET_STRING_PROP(kDriverProperty, driver);
+  LOG_INFO("Driver package: %s", driver.c_str());
+  if (driver.empty()) {
+    LOG_ERR("No driver package set");
+    std::abort();
+  }
+
+  for (char const& c : driver) {
+    if (c == ' ' || c == '"' || c == '\'') {
+      LOG_ERR("Invalid driver package: %s", driver.c_str());
+      std::abort();
+    }
+  }
+  return driver;
+}
+
+std::string getApkPath(const std::string& driver) {
+  std::string cmd = "pm path " + driver;
+  FILE* fp = popen(cmd.c_str(), "r");
+  if (fp == nullptr) {
+    LOG_ERR("Failed to run '%s'", cmd.c_str());
+    std::abort();
+  }
+
+  char buffer[1024] = {0};
+  if (fgets(buffer, sizeof(buffer) - 1, fp) == nullptr) {
+    LOG_ERR("Failed to read from pm path");
+    std::abort();
+  }
+  pclose(fp);
+
+  if (!strcmp("package:", buffer)) {
+    LOG_ERR("Unrecognized pm path output: %s", buffer);
+    std::abort();
+  }
+
+  // Remove the trailing '\n'.
+  buffer[strcspn(buffer, "\n")] = 0;
+
+  std::string path(&buffer[8]);
+  LOG_INFO("Driver package path: %s", path.c_str());
+  return path;
+}
+
+FN_PTR loadLibrary(const std::string& path, const char* lib) {
+  char *error;
+
+  std::string so = path + "!" + lib;
+  LOG_INFO("Trying %s", so.c_str());
+  void *handle = dlopen(so.c_str(), RTLD_GLOBAL);
+  if ((error = dlerror()) != nullptr || handle == nullptr) {
+    LOG_WARN("Error loading lib: %s", error);
+    return nullptr;
+  }
+
+  FN_PTR startFunc = (FN_PTR)dlsym(handle, "start");
```

Comment:
Is the starting function actually called "start". That seems ... uninformative.

Reply:
My comment exactly. Waiting on a decision in the design doc. If it gets changed, will update the code accordingly.